### PR TITLE
Fixed #23861 -- Added an API to deprecate model fields.

### PR DESCRIPTION
**Pull request by timgraham**
https://code.djangoproject.com/ticket/23861

### CODE REVIEW

File: django/db/models/fields/__init__.py [hunk 1]
```diff
@@ -290,6 +293,30 @@ def _check_null_allowed_for_primary_keys(self):
     def _check_backend_specific_checks(self, **kwargs):
         return connection.validation.check_field(self, **kwargs)
 
+    def _check_migration_availability(self):
+        if self.removed_for_migrations:
+            return [
+                checks.Error(
+                    '%s has been removed except for support in historical '
```
Comment (MarkusH) [inline django/db/models/fields/__init__.py]: You are not using `removed_for_migrations.get('message')` as a primary source here.

File: django/db/models/fields/__init__.py [hunk 2]
```diff
@@ -290,6 +293,30 @@ def _check_null_allowed_for_primary_keys(self):
...
+                    hint=self.removed_for_migrations.get('hint'),
+                    obj=self,
+                    id=self.removed_for_migrations.get('id', 'fields.EXXX'),
+                )
+            ]
+        elif self.deprecated_for_migrations:
+            return [
+                checks.Warning(
+                    self.deprecated_for_migrations.get(
+                        'message', '%s has been deprecated.' % self.__class__.__name__
```
Comment (MarkusH) [inline django/db/models/fields/__init__.py]: Can we wrap this line after the comma.

File: django/db/models/fields/__init__.py [hunk 3]
```diff
@@ -113,6 +113,8 @@ class Field(RegisterLookupMixin):
                              "%(date_field_label)s %(lookup_type)s."),
     }
     class_lookups = default_lookups.copy()
+    deprecated_for_migrations = False
```
Comment (MarkusH) [inline django/db/models/fields/__init__.py]: I'm still unsure if the names are good, but I don't have an idea for better ones.

Comment (berkerpeksag) [inline django/db/models/fields/__init__.py]: So, values of deprecated_for_migrations can be True, False and dict, right? I think this is a bit complicated. I'd set the default value to None and only accept a dict.

Comment (MarkusH) [inline django/db/models/fields/__init__.py]: Good point. Using `None` seems to make more sense, now that the variables hold more information than just yes/now.

File: tests/check_framework/test_model_field_deprecation.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+                hint='Use something else.',
+                obj=Model._meta.get_field('name'),
+                id='fields.W999',
+            )
+        ])
+
+
+class TestRemovedField(SimpleTestCase):
+    def test_default_details(self):
+        class DeprecatedField(models.Field):
```
Comment (MarkusH) [inline tests/check_framework/test_model_field_deprecation.py]: Could you rename the field to `RemovedField` given the test case name.

Comment (timgraham) [inline tests/check_framework/test_model_field_deprecation.py]: I thought about this, but then I thought, "well you don't rename a field when you remove it". Maybe I should just use something generic like `MyField` for both cases.

Comment (MarkusH) [inline tests/check_framework/test_model_field_deprecation.py]: Makes sense. Lets go with `MyField` then.

**Pull request merged**